staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -929,11 +930,33 @@ func (r *Request) newHTTPRequest(ctx context.Context) (*http.Request, error) {
 	if err != nil {
 		return nil, err
 	}
-	req = req.WithContext(ctx)
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), newDNSMetricsTrace(ctx)))
 	req.Header = r.headers
 	return req, nil
 }
 
+// newDNSMetricsTrace returns an HTTP trace that tracks time spend on DNS lookups per network/host.
+// This metric is available in client as "rest_client_dns_resolution_duration_seconds".
+func newDNSMetricsTrace(ctx context.Context) *httptrace.ClientTrace {
+	type dnsMetric struct {
+		start time.Time
+		host  string
+		sync.Mutex
+	}
+	dns := &dnsMetric{}
+	return &httptrace.ClientTrace{
+		DNSStart: func(info httptrace.DNSStartInfo) {
+			dns.Lock()
```

Comment:
I don't know if Done is guaranteed to happen for all Starts, do we want to lock the whole transaction or just avoid races accessing the variable?

Maybe we can use atomic instead of a lock https://stackoverflow.com/questions/50423209/how-to-convert-time-time-variable-to-atomic-in-go

Reply:
Added lock for the entire Start and Done.